### PR TITLE
[inductor] Enable multilayer reductions with dynamic shapes

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106581
* __->__ #106747
* #106870
* #106626

Currently multilayer reduction (aka split reductions) are only used with static
shapes which results in worse performance and accuracy when dynamic shapes are
enabled. Instead, this only requires that the shape has a hint value.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -604,9 +607,9 @@ def _is_static(x):
                 "var_unnormalized",
             }
             and config.split_reductions
-            and all(_is_static(r) for r in ranges)
-            and all(_is_static(r) for r in reduction_ranges)
-            and _is_static(reduction_numel)
+            # We don't support unbacked symints
+            and _is_static(reduction_numel_hint)
+            and _is_static(numel_hint)
```
Comment (lezcano) [inline torch/_inductor/ir.py:612]: Can it be the case that `numel` is static but `reduction_numel` is not?

Comment (lezcano) [reply torch/_inductor/ir.py:612]: In fact, we probably just want `_is_static(reduction_numel_hint)` for the general case. Then, if we also have `numel_hint` static even better, but it's not 100% necessary for the main optimisation, I don't think.

Comment (peterbell10) [reply torch/_inductor/ir.py:612]: > Can it be the case that `numel` is static but `reduction_numel` is not?

`numel` here is actually the number of output elements, so doesn't include the reduced dimensions.

> n fact, we probably just want `_is_static(reduction_numel_hint)` for the general case. Then, if we also have `numel_hint` static even better, but it's not 100% necessary for the main optimisation, I don't think.

`numel_hint` is used in conditionals like `numel_hint >= num_sm * 2 * 32` when deciding on the numbers of splits, so we need to have a concrete value.

Comment (peterbell10) [reply torch/_inductor/ir.py:612]: btw as a further step we could use `bound_sympy` to deal with unbacked SymInts but thats more than I need at this point to get cumsum working.

Comment (lezcano) [reply torch/_inductor/ir.py:612]: `bound_sympy` does not currently deal with unbacked SymInts, but may be able to do something in some cases when https://github.com/pytorch/pytorch/pull/106568 is merged.

**Comment by peterbell10** [comment]
I'm seeing geomean speedups in cudagraphs_dynamic for all benchmarks: 12% in torchbench, 17% in huggingface and 15% in Timm models.

https://hud.pytorch.org/benchmark/compilers?startTime=Wed%2C%2002%20Aug%202023%2013%3A57%3A39%20GMT&stopTime=Wed%2C%2009%20Aug%202023%2013%3A57%3A39%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/peterbell10/593/orig&lCommit=6d76b15034b7200b0a12206b050130fda70b21b0&rBranch=main&rCommit=416bf4e3e7c5f90bbc306e55501881b4f8bde981

**Comment by peterbell10** [comment]
I'm seeing geomean speedups in cudagraphs_dynamic for all benchmarks: 12% in torchbench, 17% in huggingface and 15% in Timm models.

https://hud.pytorch.org/benchmark/compilers?startTime=Wed%2C%2002%20Aug%202023%2013%3A57%3A39%20GMT&stopTime=Wed%2C%2009%20Aug%202023%2013%3A57%3A39%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/peterbell10/593/orig&lCommit=6d76b15034b7200b0a12206b050130fda70b21b0&rBranch=main&rCommit=416bf4e3e7c5f90bbc306e55501881b4f8bde981

**Review by lezcano — APPROVED** [review]
Great stuff! Perhaps also open an issue to follow up on the `bound_sympy` point?

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -595,6 +595,9 @@ def num_splits(
         def _is_static(x):
             return isinstance(x, (int, sympy.Integer))
 
+        reduction_numel_hint = V.graph.sizevars.symbolic_hint(reduction_numel)
+        numel_hint = V.graph.sizevars.symbolic_hint(sympy_product(ranges))
+
```
Comment (eellison) [inline torch/_inductor/ir.py:600]: Maybe as a follow up, add guards based on the heuristics here: https://github.com/pytorch/pytorch/blob/27705b3b40220d8f299f607eb4dafdd77484be7a/torch/_inductor/ir.py#L624

**Comment by shunting314** [comment]
I vaguely remember Voz has something similar like this but did not land? I guess the reasons is, the split reduction based on hint may not work well when the dynamic dimension change later. cc @voznesenskym  @ezyang

**Comment by peterbell10** [comment]
I couldn't find Voz's PR but with my own testing I'm seeing a pretty reasonable trade-off. 

When close to the hint size, `x.sum()` is significantly faster when split, up to 10x for very large tensors. Then if the runtime size is tiny the split kernel it's only ~1.5x slower than the non-split kernel. On the other hand if the runtime size is larger than the hint then it's still a huge win. So, this only presents an issue if the hint grossly overstates the actual runtime sizes.

**Comment by peterbell10** [comment]
Yes I agree this has no effect when the runtime size is larger than the hint. My impression was that guarding purely for performance reasons was discouraged, though maybe the accuracy implications would change that. I'm not sure, maybe @voznesenskym could weight in as he removed the old `maybe_guard` performance guards.

**Comment by ezyang** [comment]
We really need some way to indicate "please use this alternate size as the hint when compiling the dynamic shape". This is related https://github.com/pytorch/pytorch/issues/105634#issuecomment-1645624378
